modules/database/src/std/rec/mbboDirectRecord.c

```
@@ -255,60 +243,41 @@ static long special(DBADDR *paddr, int after)
         return 0;
     }
 
-    if (!after)
-        return 0;
-
-    switch (paddr->special) {
-    case SPC_MOD:   /* Bn field modified */
-        if (prec->omsl == menuOmslsupervisory) {
-            /* Adjust VAL corresponding to the bit changed */
-            epicsUInt8 *pBn = (epicsUInt8 *) paddr->pfield;
-            epicsUInt32 bit = 1 << (pBn - &prec->b0);
-
-            if (*pBn)
-                prec->val |= bit;
-            else
-                prec->val &= ~bit;
-
-            prec->udf = FALSE;
-            convert(prec);
+    if(after==0 && fieldIndex >= mbboDirectRecordB0 && fieldIndex <= mbboDirectRecordB1F) {
+        if(prec->omsl == menuOmslclosed_loop) {
+            /* To avoid confusion, reject changes to bit fields while in closed loop.
+             * Not a 100% solution as confusion can still arise if dset overwrites VAL.
+             */
+            return S_db_noMod;
```

Noting another behavior change, `special()` never rejected puts before.

---

I did most of my testing without this case.  It can be removed with few consequences.  The result would be that attempts to change B* while in closed loop will "flicker", with two updates in quick succession.  Of course the same would now happen if a device support overwrites VAL, which can't be detected ahead of time.

---

The situation that originally concerned me here was where someone had set OMSL to `closed_loop` but hadn't actually set DOL. As long as UDF is FALSE writing to the B* fields previously succeeded, but didn't make VAL change at all, then later flipping OMSL rewrote the B* fields from VAL. I agree the old behavior is confusing, so I will document the change, although I can see us getting complaints since rejecting a put (instead of just ignoring it) may trigger a pop-up on a GUI display whenever a user flips a bit for a record that's in the wrong mode, and we aren't providing any way to prevent that.